tls/_common/prf.py

```diff
@@ -0,0 +1,69 @@
+# This file is dual licensed under the terms of the Apache License, Version
+# 2.0, and the BSD License. See the LICENSE file in the root of this repository
+# for complete details.
+
+from __future__ import absolute_import, division, print_function
+
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.hmac import HMAC
+
+
+def _p_hash(hash_algorithm, secret, seed, output_length):
+    """
+    A seed expansion function that uses a single hash function to expand a
+    secret and seed into the number of bytes specified by output_length.
+    """
+    result = b""
+    i = 1
+    while len(result) < output_length:
+        h = HMAC(secret, hash_algorithm, default_backend())
+        h.update(_a(secret, hash_algorithm, i, seed))
+        h.update(seed)
+        result += h.finalize()
```

Comment:
So I ran these:

```
~ ❯❯❯ python -m timeit -s 'byte_generator = (b"foo" for _ in range(1000000))' -- "x = b''" "for chunk in byte_generator:" "    x += chunk"
10 loops, best of 3: 17.6 msec per loop
~ ❯❯❯ python -m timeit -s 'byte_generator = (b"foo" for _ in range(1000000))' -- "x = b''" "for chunk in byte_generator:" "    x += chunk"
10 loops, best of 3: 27.7 msec per loop
~ ❯❯❯ python -m timeit -s 'byte_generator = (b"foo" for _ in range(1000000))' -- "x = bytearray()" "for chunk in byte_generator:" "    x.extend(chunk)"
100000 loops, best of 3: 1.98 usec per loop
~ ❯❯❯ python -m timeit -s 'byte_generator = (b"foo" for _ in range(1000000))' -- "x = bytearray()" "for chunk in byte_generator:" "    x.extend(chunk)"
1000000 loops, best of 3: 0.345 usec per loop
~ ❯❯❯ python -m timeit -s 'byte_generator = (b"foo" for _ in range(1000000))' -- "x = []" "for chunk in byte_generator:" "    x.append(chunk)" "b''.join(x)"
1000000 loops, best of 3: 0.319 usec per loop
~ ❯❯❯ python -m timeit -s 'byte_generator = (b"foo" for _ in range(1000000))' -- "x = []" "for chunk in byte_generator:" "    x.append(chunk)" "b''.join(x)"
1000000 loops, best of 3: 0.309 usec per loop
```

The `-s` sets up things ahead of time to avoid hurting the timings. So we create our generator of one million chunks of `b'foo'` ahead of time, just to be safe.

As you can see, using `b'' += b''` hovers in the millisecond area while the list and bytearray approaches are in the microsecond area.

In short, this seems to confirm @Lukasa's assertion. That said, I was only really interested in the differences between `+=` with bytes objects, `bytearray.extend` and `list.append` with a `str.join` at the end.

What your results say to me @ashfall, is that what *probably* dominates the run time of the `_p_hash` function is `cryptography`, not how we build up the result. This makes me happy with the code as it is, and hopefully demonstrates to @markrwilliams that we don't need to refactor this function further.